crates/federated-server/src/graph/loader/object_storage.rs

```diff
@@ -119,26 +119,30 @@ impl ObjectStorageUpdater {
             .build()
             .map_err(|e| crate::Error::InternalError(e.to_string()))?;
 
-        let object_storage_host = object_storage_host();
+        let mut object_storage_url: url::Url = object_storage_host().parse().map_err(|err| {
+            crate::Error::FetcherConfigError(format!("The configured object storage host is not a valid URL: {err}"))
+        })?;
 
-        let object_storage_url = match graph_ref {
+        object_storage_url.set_path(&match graph_ref {
             GraphRef::LatestProductionVersion { graph_slug } => {
-                format!("{object_storage_host}/graphs/{graph_slug}")
+                format!("/graphs/{graph_slug}")
             }
             GraphRef::LatestVersion {
                 graph_slug,
                 branch_name,
-            } => format!("{object_storage_host}/graphs/{graph_slug}/branch/{branch_name}"),
+            } => {
+                let branch_name = url::form_urlencoded::byte_serialize(branch_name.as_bytes()).collect::<String>();
```

Comment:
Won't we have the same problem with graph names?

Reply:
Thought so too, but we validate that these only have alphanumeric characters and hyphens in the api, so they should be fine: https://github.com/grafbase/api/blob/47a6be29e3e5e76912dd603ccfc784ce3b810bb3/crates/domain/src/validation.rs